agent/src/scheduler/mod.rs

```diff
@@ -0,0 +1,508 @@
+/// across the entire fleet of runners, and jobs can have requirements that act as constraints when
+/// selecting runners.
+///
+use ssri::Integrity;
+use thiserror::Error;
+use uuid::Uuid;
+
+/// The maximum amount of time that a runner should be given to run the job without checking back in
+/// with the scheduler; after this amount of time passes, the current run will be considered failed
+/// and the job will either be retried or abandoned, depending on how many prior run attempts have
+/// been made (see MAX_JOB_ATTEMPTS).
+const MAX_JOB_DURATION: Duration = Duration::from_secs(60);
+
+/// The maximum number of times to try running a job that has previously timed out.
+const MAX_JOB_ATTEMPTS: u8 = 3;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ServalSchedulerError {
+    #[error("The attempted operation is not valid for the current state of the job")]
+    InvalidOperationForJobState,
+
+    #[error("The given job could not be located")]
+    JobNotFound,
+}
+
+/// Represents the state of a job that the scheduler is currently taking care of; transient
+/// information that is only relevant to a job during a particular part of its lifetime should live
+/// within one of these enum values.
+#[derive(Debug, PartialEq)]
+enum ScheduledJobState {
+    /// The job is waiting to be assigned to a runner.
+    Unassigned,
+    /// The job has been assigned to a runner and has a deadline, at which point it will be marked
+    /// as failed and either move back to Unassignd or marked as Failed.
+    InProgress { runner: Uuid, deadline: SystemTime },
+    ///  The job was completed successfully.
+    Completed {
+        runner: Uuid,
+        completion_time: SystemTime,
+        output: Option<Integrity>,
+    },
+    /// The job failed to complete; either it timed out MAX_JOB_ATTEMPTS times, or it was explicitly
+    /// marked as failed by a runner.
+    Failed {
+        runner: Uuid,
+        failure_time: SystemTime,
+        output: Option<Integrity>,
+    },
+}
+
+/// The priority with which a given job should be scheduled.
+#[derive(Debug, Default, PartialEq)]
+enum ScheduledJobPriority {
+    /// This job should take priority over all jobs of lower priority.
+    Emergency = 0,
+    /// This job should take priority over all non-emergency jobs.
+    HighPriority = 1,
+    /// This job should be scheduled with a normal amount of priority.
+    #[default]
+    Normal = 2,
+    /// This job should only be scheduled if there is nothing more important to do.
+    LowPriority = 3,
+}
+
+/// Represents a constraint on the runner that can run a job.
+#[derive(Debug, PartialEq)]
+enum ScheduledJobRequirement {
+    /// Requires that the runner has the given extension available.
+    Extension(String),
+    /// Requires that the runner has `/proc` (e.g. is Linux)
+    Proc,
+}
+
+/// Represents the kind of invocation this job would like to receive.
+#[derive(Debug, PartialEq)]
+enum ScheduledJobKind {
```

Comment:
These are referred to as "Distribution Models" in the spec, and as "Kind" here. I don't like either of these terms. @ceejbot I am begging you to help me come up with a good name for this concept.